### ISSUE TITLE
Jupyter kernel: I/O operation on closed file

### ISSUE THREAD

**Issue by kalekundert** [xonsh/xonsh#2099]
I had to make an awful hack to get the Jupyter Xonsh kernel working, I'm hoping somebody can either find a real solution or just tell me that I'm doing something wrong.

When I enter a python-mode command (e.g. "1+1") into jupyter, everything works as expected.  When I enter a shell-mode command (e.g. "ls"), I get the following stack trace (in the notebook server, not the client):

```
[IPKernelApp] ERROR | Exception in message handler:
Traceback (most recent call last):
  File "/home/kale/.local/lib/python3.6/site-packages/xonsh/base_shell.py", line 319, in default
    run_compiled_code(code, self.ctx, None, 'single')
  File "/home/kale/.local/lib/python3.6/site-packages/xonsh/codecache.py", line 65, in run_compiled_code
    func(code, glb, loc)
  File "<xonsh-code>", line 1, in <module>
  File "/home/kale/.local/lib/python3.6/site-packages/ipykernel/displayhook.py", line 34, in __call__
    sys.stdout.flush()
  File "/home/kale/.local/lib/python3.6/site-packages/xonsh/base_shell.py", line 170, in flush
    self.std.flush()
  File "/home/kale/.local/lib/python3.6/tempfile.py", line 709, in flush
    self._file.flush()
ValueError: I/O operation on closed file.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/kale/.local/lib/python3.6/site-packages/ipykernel/kernelbase.py", line 228, in dispatch_shell
    handler(stream, idents, msg)
  File "/home/kale/.local/lib/python3.6/site-packages/ipykernel/kernelbase.py", line 390, in execute_request
    user_expressions, allow_stdin)
  File "/home/kale/.local/lib/python3.6/site-packages/xonsh/jupyter_kernel.py", line 55, in do_execute
    shell.default(code)
  File "/home/kale/.local/lib/python3.6/site-packages/xonsh/base_shell.py", line 328, in default
    print_exception()
  File "/home/kale/.local/lib/python3.6/site-packages/xonsh/tools.py", line 630, in print_exception
    sys.stderr.write('xonsh: For full traceback set: '
  File "/home/kale/.local/lib/python3.6/site-packages/xonsh/base_shell.py", line 165, in write
    self.std.write(std_s)
  File "/home/kale/.local/lib/python3.6/tempfile.py", line 766, in write
    rv = file.write(s)
ValueError: I/O operation on closed file.
```

This is using python3.6 and xonsh 0.5.2.  I'm not sure how relevant that is.

It looks like the problem is that ``XonshKernel.do_execute()`` opens up temporary files to capture stdout and stderr, which it will later try to report to jupyter.  It then calls ``BaseShell.default()`` to handle the command entered by the user.  ``BaseShell.default()`` also captures stdout and stderr, this time using a ``Tee`` object, and closes the files when it's done doing its thing.  When control returns to ``XonshKernel.do_execute()``, it tries to read from its temporary files and crashes because they've been closed.

I was able to work-around the problem just by commenting out all the lines involving ``tee`` in ``BaseShell.default()``, but that's obviously a hack and it looks like it'll ruin the history somehow.  I also don't understand why this isn't a problem for python-mode commands, or for anyone else (as far as I can tell).

Anways, I'd like to get some feedback on what's happening and what some good solutions might be.  I'd be more than happy to submit a pull request addressing this if there's a consensus about what that pull request should look like.

**Comment by gforsyth** [comment]
Hey @kalekundert, thanks for reporting!  And sorry for the trouble.  I can reproduce this with Python 3.5, too.  I'm going to try to dig in on this a little -- it looks like we haven't updated the jupyter kernel since the major subprocess refactor, so that is likely why things aren't working. I think your diagnosis is correct and it will definitely screw up history. 

And I think there's a similar jupyter kernel issue reported, so you aren't alone.

**Comment by gforsyth** [comment]
I believe this is a duplicate of #2044

**Comment by kalekundert** [comment]
If you can give me a general description of what needs to change, I might be able to do it.  There were too many moving parts for me to want to get into it by myself, but with some direction I think I could manage it.

**Comment by gforsyth** [comment]
Hey @kalekundert -- sorry to leave you hanging on this. I'm honestly not sure what needs to change, beyond the obvious fact that we can't close the same file twice.

**Comment by scopatz** [comment]
Potential fix is in #2113

**Comment by scopatz** [comment]
It seems to work for me locally.
